doc/topics/development/modules/index.rst

```diff
@@ -0,0 +1,279 @@
+===============
+Modular Systems
+===============
+
+When first working with Salt, it is not always clear where all of the modular
+components are and what they do. Salt comes loaded with more modular systems
+than many users are aware of, making Salt very easy to extend in many places.
+
+The most commonly used modular systems are execution modules and states. But
+the modular systems extend well beyond the more easily exposed components
+and are often added to Salt to make the complete system more flexible.
+
+Loading Modules
+===============
+
+Modules come primarily from several sources:
+
+* The Salt package itself
+* The Salt File Server
+* The extmods directory
+* Secondary packages installed
+
+Using one source to override another is not supported.
+
+The Salt Package
+----------------
+
+Salt itself ships with a large number of modules. These are part of the Salt
+package itself and don't require the user to do anything to use them. (Although
+a number of them have additional dependencies and/or configuration.)
+
+The Salt File Server
+--------------------
+
+The user may add modules by simply placing them in special directories in their
+:ref:`fileserver <file-server>`.
+
+These will be eventually picked up by minions, or users can use
+the ``saltutil.sync_*`` :py:mod:`execution functions <salt.modules.saltutil>`
+and :py:mod:`runner functions <salt.runners.saltutil>` to force it immediately.
+
+The name of the directory inside of the file server is the directory name
+prepended by ``_``.
+
+Using saltenvs besides ``base`` may not work in all contexts.
+
+The extmods Directory
+---------------------
+
+Any files places in the directory set by ``extension_modules`` settings
+(:conf_minion:`minion <extension_modules>`,
+:conf_master:`master <extension_modules>`, default
+``/var/cache/salt/*/extmods``) can also be loaded as modules. Note that these
+directories are also used by the ``saltutil.sync_*`` functions (mentioned
+above).
+
+Secondary Packages
+------------------
+
+Third-party packages may also add modules to Salt if they are installed in the
+same system and Python environment as the Salt Minion or Master.
+
+This is done via setuptools entry points:
+
+.. code-block:: python
```

Comment:
This needs a newline below it or the documentation will not build.

Reply:
Thanks! Handled.